source/common/network/proxy_protocol.cc

```diff
@@ -61,17 +60,47 @@ void ProxyProtocol::ActiveConnection::onReadWorker() {
     return;
   }
 
-  if (proxy_line.find(PROXY_TCP4) != 0) {
+  // Parse proxy protocol line with format: PROXY TCP4/TCP6 SOURCE_ADDRESS DESTINATION_ADDRESS
+  // SOURCE_PORT DESTINATION_PORT.
+  auto line_parts = StringUtil::split(proxy_line, " ", true);
+
+  if (line_parts.size() != 6 || line_parts[0] != "PROXY") {
+    throw EnvoyException("failed to read proxy protocol");
+  }
+
+  Address::IpVersion protocol_version;
+  if (line_parts[1] == "TCP4") {
+    protocol_version = Address::IpVersion::v4;
+  } else if (line_parts[1] == "TCP6") {
+    protocol_version = Address::IpVersion::v6;
+  } else {
     throw EnvoyException("failed to read proxy protocol");
   }
 
-  size_t index = proxy_line.find(" ", PROXY_TCP4.size());
-  if (index == std::string::npos) {
+  // Error check the source and destination fields. Currently the source port, destination address,
+  // and destination port fields are ignored. Remote address refers to the source address.
+  Address::InstanceConstSharedPtr remote_address = Utility::parseInternetAddress(line_parts[2]);
+  Address::InstanceConstSharedPtr destination_address =
+      Utility::parseInternetAddress(line_parts[3]);
+  auto remote_version = remote_address->ip()->version();
+  auto destination_version = destination_address->ip()->version();
+  if (remote_version != protocol_version || destination_version != protocol_version) {
     throw EnvoyException("failed to read proxy protocol");
   }
 
-  size_t addr_len = index - PROXY_TCP4.size();
-  std::string remote_address = proxy_line.substr(PROXY_TCP4.size(), addr_len);
+  uint64_t remote_port, destination_port;
+  try {
+    remote_port = std::stol(line_parts[4]);
+    destination_port = std::stol(line_parts[5]);
+  } catch (const std::invalid_argument& ia) {
+    throw EnvoyException(ia.what());
+  } catch (const std::out_of_range& ex) {
+    throw EnvoyException(ex.what());
+  }
+  if (long(remote_port) < 0 || long(destination_port) < 0) {
```

Comment:
It's a bit unexpected that the original test has source port 66776 which is > 65536. The protocol specifies that these values are the TCP source/destination port. Adding the extra constraint would break this test.